core/commonMain/src/implementations/immutableSet/PersistentHashSet.kt

```diff
@@ -21,6 +22,13 @@ internal class PersistentHashSet<E>(internal val node: TrieNode<E>,
     }
 
     override fun addAll(elements: Collection<E>): PersistentSet<E> {
+        if (elements is PersistentHashSet) {
+            val intersectionCountRef = DeltaCounter()
+            val newNode = node.addAll(elements.node, 0, intersectionCountRef)
+            if (node === newNode) return this
+            if (elements.node === newNode) return elements
+            return PersistentHashSet(newNode, size + elements.size - intersectionCountRef.count)
+        }
         return this.mutate { it.addAll(elements) }
```

Comment:
Does optimizing `addAll` for `PersistentHashSet` significantly improves performance now that `builder.addAll` is optimized?
Looking through the implementations they seem almost identical. If the performance impact is not significant, optimizing only for builder is more favourable.

Reply:
Sacrificing a throwaway builder to avoid duplication of big non-trivial code seems like a bargain.

Reply:
Fair point. If we can work out the current issues of builder implementation, having only one method is a good goal to match.

Reply:
Especially considering that `retainAll`, `removeAll` and `containsAll` may also be optimized (the last one having huge speedup as per my experiments) and having two versions of each of these methods would be hell.